### PR TITLE
Add a new wasi_nn backend related to burn.rs

### PR DESCRIPTION
**Pull request by CaptainVincent**
As our show cases, currently two models are supported: 

SqueezeNet and Whisper.

But the plugin currently supports only one model at a time.

ex.

```bash
cmake -GNinja -Bbuild -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_BUILD_AOT_RUNTIME=OFF -DWASMEDGE_PLUGIN_WASI_NN_RUST_MODEL=whisper

or 

cmake -GNinja -Bbuild -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_BUILD_AOT_RUNTIME=OFF -DWASMEDGE_PLUGIN_WASI_NN_RUST_MODEL=squeezenet
```

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review agent](https://github.com/flows-network/github-pr-review/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

**Comment by CaptainVincent** [comment]
The missing wasi-nn headers in the wasiNNTest for the burn plugin require this part to be rewritten, which will take some time. Considering the priority, I keep only the release CI for now.

File: plugins/wasi_nn_rust/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+if(CMAKE_BUILD_TYPE STREQUAL "Debug")
```
Comment (q82419) [inline plugins/wasi_nn_rust/CMakeLists.txt:1]: Add the liscense.

File: plugins/wasi_nn_rust/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+message(STATUS "WasmEdge Wasi-NN Rust plugin model: ${WASMEDGE_PLUGIN_WASI_NN_RUST_MODEL}")
+set(CARGO_FEATURES "--features=${WASMEDGE_PLUGIN_WASI_NN_RUST_MODEL}")
+
+set(RS_SO ${CMAKE_CURRENT_BINARY_DIR}/${TARGET_DIR}/libwasmedgePluginWasiNN${CMAKE_SHARED_LIBRARY_SUFFIX})
+
+set(WASMEDGE_LIB_DIR ${CMAKE_CURRENT_BINARY_DIR}/../../lib/api)
+
+add_custom_target(wasi_nn_rust ALL
+  COMMAND WASMEDGE_LIB_DIR=${WASMEDGE_LIB_DIR} LD_LIBARAY_PATH=${WASMEDGE_LIB_DIR} CARGO_TARGET_DIR=${CMAKE_CURRENT_BINARY_DIR} ${CARGO_CMD} ${CARGO_FEATURES}
+  COMMAND cp ${RS_SO} ${CMAKE_CURRENT_BINARY_DIR}
```
Comment (q82419) [inline plugins/wasi_nn_rust/CMakeLists.txt:18]: Use `${CMAKE_COMMAND} -E copy` here.

File: plugins/wasi_nn_rust/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  COMMAND rm -rf ${CMAKE_CURRENT_BINARY_DIR}/${TARGET_DIR}
```
Comment (q82419) [inline plugins/wasi_nn_rust/CMakeLists.txt:19]: Use `${CMAKE_COMMAND} -E remove` here.

File: plugins/wasi_nn_rust/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,38 @@
+[package]
+name = "wasi_nn_rust"
+version = "0.0.1"
+edition = "2021"
```
Comment (q82419) [inline plugins/wasi_nn_rust/Cargo.toml:4]: Should we use `2023` here?

File: plugins/wasi_nn_rust/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,382 @@
+mod helper;
+mod models;
+
+pub enum ErrNo {
```
Comment (dm4) [inline plugins/wasi_nn_rust/src/lib.rs:4]: Is it possible to re-use [error.rs](https://github.com/second-state/wasmedge-wasi-nn/blob/ggml/rust/src/error.rs) from wasmedge-wasi-nn?

File: plugins/wasi_nn_rust/CMakeLists.txt (hunk 1 above)
Comment (CaptainVincent) [reply plugins/wasi_nn_rust/CMakeLists.txt:4]: Done.

File: plugins/wasi_nn_rust/CMakeLists.txt (hunk 3 above)
Comment (CaptainVincent) [reply plugins/wasi_nn_rust/CMakeLists.txt]: Done.

File: plugins/wasi_nn_rust/CMakeLists.txt (hunk 2 above)
Comment (CaptainVincent) [reply plugins/wasi_nn_rust/CMakeLists.txt]: Done.

File: plugins/wasi_nn_rust/Cargo.toml (hunk 4 above)
Comment (CaptainVincent) [reply plugins/wasi_nn_rust/Cargo.toml:4]: It seems that the 2023 edition is not available. The latest edition of Rust that can be used is 2024, but it is currently still in [beta](https://doc.rust-lang.org/edition-guide/rust-2024/index.html). Here, the previous usable version, 2021, is kept to resolve this issue.

File: plugins/wasi_nn_rust/src/lib.rs (hunk 5 above)
Comment (CaptainVincent) [reply plugins/wasi_nn_rust/src/lib.rs:4]: I tried it out, but I found that there is no explicit method to convert it to i32. Therefore, I still need to manually write the conversion function in the plugin layer (the code on both sides will still decoupled in the future). And it also lacks a Success enum. Given these reasons, it is less concise than declaring a new enum directly, so I prefer to keep the existing approach.

Comment (dm4) [reply plugins/wasi_nn_rust/src/lib.rs:4]: Sounds reasonable.

**Review by q82419 — APPROVED** [review]

**Pull request merged**